### ISSUE TITLE
add "good_enough" or "satisfactory" confidence

### ISSUE THREAD

**Issue by hqhoang** [baudneo/neo-mlapi#1]
(this should go into the mlapi project but your repo doesn't allow submitting issues there, sorry)

I'm facing a situation where a car is about to turn into my driveway, but its headlight triggers motion detection too early, resulting in the first 100-200 frames have nothing to detect. One solution is to keep checking the frames in interval, e.g. 10, 20, 30, 40, ... until we get something, but then the "most" strategy would require checking all the available frames to find the best, making the detection/announcement quite late.

To balance out between accuracy vs immediacy, I'm proposing a "satisfactory_confidence" param.

For example, I set `min_conf = 0.4`, but set `satisfactory_conf = 0.8`. The detection could go through the frames sequentially, but instead of waiting to get all the frames and picking out the best, if it hits a frame with 0.8+ conf, it would stop and report right away. If it doesn't hit any frame with 0.8+, then it'd pick the best like usual, something in the range [0.4-0.8).

That way, in the scenario above, the car could wait on the street and the detection will keep going through the frames. But as soon as the car enters the driveway, the detection would likely detect the car with a satisfactory_conf of 0.8+, thus, stop and report the detection right away instead of waiting for the event to be over. 

Like the other params, this should be overridable per monitor, and optionally per object type.

**Comment by baudneo** [comment]
This is an interesting idea. I'll take a look at implementing but won't promise it will stay. 

What happens when you use the 'first' strategy? Is there a post-filtered detection that is causing it to end detections early? The only reason I ask is it see if this may be achievable with current filters. However, I can see how a filter like this may be useful for other scenarios though, so I will give it implementation in a new branch based on the pull_req branch.

**Comment by hqhoang** [comment]
I started with "first" strategy, but encountered too many false positives with a low min_conf, and missed too many with a high min_conf.

**Comment by baudneo** [comment]
There are several things to consider for this filter if it hits. 
1. Does strategy matter? i.e. should the 'good enough' option only be applied on 'most' or 'first'?
2. When a frame hits the 'good enough' confidence, is it accepted as a hit unconditionally or does it continue to be processed through the filtering function? i.e. X hits 80% but it is not inside of the specified zone or does not have enough pixels inside of the zone, is it still a hit? This is important due to the way the filters are applied.
3. Should this be specifically defined per zone instead of a blanket option?

2 is most important to figure out, I think. The way that it is handled will dictate how to implement the logic.

Here are some logs of what I have put together already, just need to iron out a few things.
```
01/26/22 02:45:03.623823 zm_mlapi[11211] DBG2 object:36 [yolo:portalock: 'pyzm_uid1000_GPU_lock' already acquired]
  01/26/22 02:45:03.626239 zm_mlapi[11211] DBG1 yolo:101 [yolo: loading model data from sequence 'DarkNet::v4 Pre-Trained']
  01/26/22 02:45:03.730207 zm_mlapi[11211] DBG1 yolo:109 [perf:yolo: 'DarkNet::v4 Pre-Trained' initialization -> loading 'yolov4.weights' took: 101.37 ms]
  01/26/22 02:45:03.730435 zm_mlapi[11211] INF yolo:122 [yolo: OpenCV (4.5.4+) fix for getUnconnectedOutLayers() API (Non nested structure - Thanks @pliablepixels !)]
  01/26/22 02:45:03.735181 zm_mlapi[11211] DBG1 yolo:142 [yolo: using GPU for detection, set CUDA/cuDNN backend and target]
  01/26/22 02:45:03.737683 zm_mlapi[11211] DBG1 yolo:200 [yolo: 'DarkNet::v4 Pre-Trained' (GPU) - input image 1920*1080 - resized by  model_width/height to: 416*416]
  01/26/22 02:45:04.588633 zm_mlapi[11211] DBG2 object:62 [yolo:portalock: released 'pyzm_uid1000_GPU_lock']
  01/26/22 02:45:05.009074 zm_mlapi[11211] DBG2 yolo:313 [perf:yolo:GPU: 'DarkNet::v4 Pre-Trained' detection took: 1268.92 ms]
  01/26/22 02:45:05.011432 zm_mlapi[11211] DBG1 yolo:322 [yolo: ['car', 'car'] -- [[542, 309, 1158, 581], [735, 184, 839, 246]] -- [0.9542021751403809, 0.31044507026672363]]
  01/26/22 02:45:05.015083 zm_mlapi[11211] DBG2 detect_sequence:1277 [detect: model: 'object' seq: 'DarkNet::v4 Pre-Trained' found 2 detections -> car, car]
  01/26/22 02:45:05.017325 zm_mlapi[11211] DBG1 detect_sequence:375 [DEBUG!>>> SEQUENCE OPTIONS min_conf = '0.6' -- min_conf_found = 'Sequence DarkNet::v4 Pre-Trained -> object_min_confidence:']
  01/26/22 02:45:05.019563 zm_mlapi[11211] DBG1 detect_sequence:400 [DEBUG!>>> ml_sequence OPTIONS conf_upper = '80%' -- conf_upper_found = 'ml_sequence object:general -> confidence_upper:']
  01/26/22 02:45:05.02181 zm_mlapi[11211] DBG1 detect_sequence:446 [>>> detected 'car (1/2)' confidence: 0.95]
  01/26/22 02:45:05.024134 zm_mlapi[11211] DBG1 detect_sequence:576 ['car (1/2)' minimum confidence found: (Sequence DarkNet::v4 Pre-Trained -> object_min_confidence:) -> '0.6']
  01/26/22 02:45:05.026316 zm_mlapi[11211] DBG1 detect_sequence:594 [
   -- _m.groups() = ('80', '%')]
  01/26/22 02:45:05.028498 zm_mlapi[11211] DBG1 detect_sequence:598 [CONFIDENCE UPPER IS EXPLICITLY % ---- _m.group(1) = '80']
  01/26/22 02:45:05.030667 zm_mlapi[11211] DBG1 detect_sequence:612 ['car (1/2)' UPPER confidence found: (ml_sequence object:general -> confidence_upper:) -> '0.8']
  01/26/22 02:45:05.032868 zm_mlapi[11211] DBG2 detect_sequence:616 [confidence: 95.42 is equal to or higher than 80.00, accepting as a match..]
  01/26/22 02:45:05.035092 zm_mlapi[11211] DBG2 detect_sequence:652 [checking if 'car (1/2)' @ [542, 309, 1158, 581] is inside polygon/zone 'parking_area' located at [(805, 200), (1897, 125), (1910, 562), (7, 594)
  ]]
  01/26/22 02:45:05.037318 zm_mlapi[11211] DBG1 detect_sequence:657 ['car (1/2)' INTERSECTS polygon/zone 'parking_area']
  01/26/22 02:45:05.039875 zm_mlapi[11211] DBG2 detect_sequence:662 ['car (1/2)' has 165910.93 pixels (99.02%) inside 'parking_area']
  01/26/22 02:45:05.042177 zm_mlapi[11211] DBG3 detect_sequence:715 [match pattern: 'car (1/2)' zone 'parking_area' has overrides->'(person|car)']
  01/26/22 02:45:05.044696 zm_mlapi[11211] DBG1 detect_sequence:772 [mpd: there are no saved detections to evaluate, skipping match past detection filter]
  01/26/22 02:45:05.046883 zm_mlapi[11211] DBG2 detect_sequence:978 [detection: 'car (1/2)' has PASSED filtering]
  01/26/22 02:45:05.04911 zm_mlapi[11211] DBG1 detect_sequence:441 [mpd: CONFIDENCE UPPER HAS BEEN HIT AND A MATCH WAS APPENDED TO THE RETURN DATA BREAKING OUT OF FILTERING LOOP!]
  01/26/22 02:45:05.05138 zm_mlapi[11211] DBG2 detect_sequence:1315 [detect:strategy: '1' filtered label: ['car'] [0.9542021751403809] ['yolo[GPU]'] [[542, 309, 1158, 581]]]
  01/26/22 02:45:05.053527 zm_mlapi[11211] DBG2 detect_sequence:1409 [perf:frame: s-103 took 4427.62 ms]
  01/26/22 02:45:05.057447 zm_mlapi[11211] DBG2 detect_sequence:1427 [detect: breaking out of frame loop as 'frame_strategy' is 'first']
  01/26/22 02:45:05.059664 zm_mlapi[11211] DBG2 object:57 [coral:portalock: already released 'pyzm_uid1000_TPU_lock']
  01/26/22 02:45:05.061861 zm_mlapi[11211] DBG2 object:57 [yolo:portalock: already released 'pyzm_uid1000_GPU_lock']
  01/26/22 02:45:05.063924 zm_mlapi[11211] DBG1 detect_sequence:1489 [perf:detect:FINAL: 'Monitor': Back Alley - MODECT (2)->'Event': 62889 -> complete detection sequence took: 4727.97 ms]
  01/26/22 02:45:05.065984 zm_mlapi[11211] DBG1 detect_sequence:1501 [Adding the current detection to the MATCH PAST DETECTION buffer!]
```

**Comment by baudneo** [comment]
Logs for work with the short cicuit being tested:

Success, not removed by any filters after a 'good enough' hit. Things seem to work in my testing so far. I will push for testing within 24 hours.
```
  01/26/22 03:40:14.620786 zm_mlapi[11921] DBG1 detect_sequence:451 [>>> detected 'car (1/2)' confidence: 0.95]
  01/26/22 03:40:14.623417 zm_mlapi[11921] DBG1 detect_sequence:581 ['car (1/2)' minimum confidence found: (Sequence DarkNet::v4 Pre-Trained -> object_min_confidence:) -> '0.6']
  01/26/22 03:40:14.625683 zm_mlapi[11921] DBG1 detect_sequence:599 [
   -- _m.groups() = ('80', '%')]
  01/26/22 03:40:14.627745 zm_mlapi[11921] DBG1 detect_sequence:603 [CONFIDENCE UPPER IS EXPLICITLY % ---- _m.group(1) = '80']
  01/26/22 03:40:14.629836 zm_mlapi[11921] DBG1 detect_sequence:617 ['car (1/2)' UPPER confidence found: (ml_sequence object:general -> confidence_upper:) -> '0.8']
  01/26/22 03:40:14.631852 zm_mlapi[11921] DBG2 detect_sequence:621 [confidence: 95.42 is equal to or higher than 80.00, accepting as a match..]
  01/26/22 03:40:14.633947 zm_mlapi[11921] DBG2 detect_sequence:657 [checking if 'car (1/2)' @ [542, 309, 1158, 581] is inside polygon/zone 'parking_area' located at [(805, 200), (1897, 125), (1910, 562), (7, 594)
  ]]
  01/26/22 03:40:14.635995 zm_mlapi[11921] DBG1 detect_sequence:662 ['car (1/2)' INTERSECTS polygon/zone 'parking_area']
  01/26/22 03:40:14.638304 zm_mlapi[11921] DBG2 detect_sequence:667 ['car (1/2)' has 165910.93 pixels (99.02%) inside 'parking_area']
  01/26/22 03:40:14.640359 zm_mlapi[11921] DBG3 detect_sequence:720 [match pattern: 'car (1/2)' zone 'parking_area' has overrides->'(person|car)']
  01/26/22 03:40:14.642521 zm_mlapi[11921] DBG1 detect_sequence:777 [mpd: there are no saved detections to evaluate, skipping match past detection filter]
  01/26/22 03:40:14.644506 zm_mlapi[11921] DBG2 detect_sequence:983 [detection: 'car (1/2)' has PASSED filtering]
  01/26/22 03:40:14.646427 zm_mlapi[11921] DBG1 detect_sequence:441 [detect: CONFIDENCE UPPER HAS BEEN HIT AND A MATCH WAS APPENDED TO THE RETURN DATA, BREAKING OUT OF FILTERING LOOP!]
  01/26/22 03:40:14.648325 zm_mlapi[11921] DBG2 detect_sequence:1322 [detect:strategy: '1' filtered label: ['car'] [0.9542021751403809] ['yolo[GPU]'] [[542, 309, 1158, 581]]]
  01/26/22 03:40:14.650239 zm_mlapi[11921] DBG3 detect_sequence:1377 [detect: SHORT CIRCUITING same_model_sequence_strategy FOR CONFIDENCE_UPPER]
  01/26/22 03:40:14.652215 zm_mlapi[11921] DBG2 detect_sequence:1432 [perf:frame: s-103 took 7993.21 ms]
  01/26/22 03:40:14.655972 zm_mlapi[11921] DBG3 detect_sequence:1456 [detect: SHORT CIRCUITING frame_strategy FOR CONFIDENCE_UPPER]
  01/26/22 03:40:14.657902 zm_mlapi[11921] DBG2 object:57 [coral:portalock: already released 'pyzm_uid1000_TPU_lock']
  01/26/22 03:40:14.659757 zm_mlapi[11921] DBG2 object:57 [yolo:portalock: already released 'pyzm_uid1000_GPU_lock']
  01/26/22 03:40:14.66156 zm_mlapi[11921] DBG1 detect_sequence:1518 [perf:detect:FINAL: 'Monitor': Back Alley - MODECT (2)->'Event': 62889 -> complete detection sequence took: 8295.42 ms]
  01/26/22 03:40:14.663331 zm_mlapi[11921] DBG1 detect_sequence:1530 [Adding the current detection to the MATCH PAST DETECTION buffer!]
```

**Comment by hqhoang** [comment]
wow, thanks much for the quick implementation! I'll check the logic and test the code when it's available.

I still have problem connecting to remote mlapi (on the same host, via port 5000), so I'm just testing with local detection only at the moment. Here's the error, I haven't tried to trace/debug it yet.
```
zmesdetect_m4[42213] DBG1 zm_detect:1047->[Traceback (most recent call last):
  File "zm_detect.py", line 1014, in main_handler
    remote_response = remote_detect(options=stream_options, args=args, route=route)
  File "zm_detect.py", line 487, in remote_detect
    kickstart: dict = g.api.cred_dump()
  File "/usr/local/lib/python3.8/dist-packages/pyzm/api.py", line 168, in cred_dump
    "access_token_datetime": self.access_token_datetime.timestamp(),
AttributeError: 'NoneType' object has no attribute 'timestamp'
]
```

**Comment by baudneo** [comment]
Oh my that's one I haven't seen yet. I will add a try block to catch what's going on. Pushing code shortly to the pyzm repo under a branch named 'good_enough' with the new filter. The name of the option will be 'confidence_upper' and '<LABEL>_confidence_upper' - currently the pre-defined 'zones' do not support this filter. You will need to add the filter to your configs (objectconfig and/or mlapiconfig).

The filters and their per label counterparts can be defined in 
- ml_sequence>general
- ml_sequence>MODEL_NAME>general
- ml_sequence>MODEL_NAME>sequences>per-sequence
- pre-defined 'zones' (except this 'good enough' filter for now)

Precedence goes in reverse order, meaning pre-defined zones takes precedence than the sequence, then the model>general and lastly is ml_seqeunce>general.

CONFIG EXAMPLE
```
monitors:

  2:
    #frame_set:  'snapshot,70,110,160,snapshot,alarm'
    object_detection_pattern: (person)
    person_max_detection_size: 60%
    #Back Alley / Car  - CLONE with Modect
    # Legacy way to define polygon and its detection pattern NOTE: _polygonzone and _polygon_zone are both valid
#    parking_area_zone_detection_pattern: (person)
#    parking_area_polygonzone: 0,496 1910,0 1910,634 0,640
    #    person_min_confidence: 0.60
    frame_set: snapshot,70,110,160,alarm,210,280,snapshot
    past_det_max_diff_area: 0.16
    car_past_det_max_diff_area: 34%
    confidence_upper: 80%
```

**Comment by baudneo** [comment]
That exception may be due to your system being too fast lol (I'm running on r710 HW), I think I need to enable inter-thread communication using locks, not certain yet until more error data.

**Comment by baudneo** [comment]
pyzm 'good_enough' branch has been pushed.

**Comment by hqhoang** [comment]
Was able to add "confidence_upper" key into objectconfig.yml and caused short-circuiting, very nice, reduces the detection time in half for a typical scenario.

Still get that same error, not very useful error information. What else do I need to help debugging this problem?

```
01/26/22 18:26:28.029493 zmesdetect_m13[6750] DBG1 api.py:178->[CRED DUMP DEBUG>>>  exception as str -> 'NoneType' object has no attribute 'timestamp']

01/26/22 18:26:28.030377 zmesdetect_m13[6750] DBG1 api.py:179->[Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/pyzm/api.py", line 171, in cred_dump
    "access_token_datetime": self.access_token_datetime.timestamp(),
AttributeError: 'NoneType' object has no attribute 'timestamp'
]

01/26/22 18:26:28.002849 zmesdetect_m13[6750] ERR api.py:177->[api: ERROR while attempting to dump current credentials]
```

Forgot to mention that I had to modify two other spots previously to avoid getting error, too.
```
diff --git a/pyzm/ZMLog.py b/pyzm/ZMLog.py
index 17a28c3..211971e 100644
--- a/pyzm/ZMLog.py
+++ b/pyzm/ZMLog.py
@@ -366,7 +366,7 @@ class ZMLog:
         if self.buffer.buffer and len(self.buffer.buffer):
             self.buffer.buffer = sorted(self.buffer.buffer, key=lambda x: x['timestamp'], reverse=True)
             for _ in range(len(self.buffer)):
-                if self.buffer:
+                if self.buffer is not None:
                     # pop it just in case it somehow gets iterated again
                     line = self.buffer.pop()
                     if line:
diff --git a/pyzm/helpers/pyzm_utils.py b/pyzm/helpers/pyzm_utils.py
index 5534284..29ec597 100644
--- a/pyzm/helpers/pyzm_utils.py
+++ b/pyzm/helpers/pyzm_utils.py
@@ -1425,8 +1425,10 @@ class LogBuffer:
         if self.buffer and len(self.buffer):
             # sort it by timestamp
             self.buffer = sorted(self.buffer, key=lambda x: x['timestamp'], reverse=True)
+
             for _ in range(len(self.buffer)):
-                line = self.buffer.pop()
+                if (len(self.buffer) > 0): 
+                    line = self.buffer.pop()
                 if line:
                     fnfl = f"{line['filename']}:{line['lineno']}"
                     print_log_string = (f"{line['timestamp']} LOG_BUFFER[{os.getpid()}] {line['display_level']} " 

```

**Comment by hqhoang** [comment]
Also, not sure if it would make any improvement in performance, but I noticed that after a detection, the confidence is checked against `min_conf` before checking against the labels. There are a lot of objects detected in a room: tv, chair, table, ... so the confidences were all checked before they're all discarded due to not matching the label filter. I guess these are simple calculation, doesn't affect performance much, maybe low priority.

**Comment by baudneo** [comment]
Can you post more of the logs previous to the exception after pulling this next commit. I am adding a few debug statements when the offending vars are set.

Are you using API token auth?

**Comment by baudneo** [comment]
> Also, not sure if it would make any improvement in performance, but I noticed that after a detection, the confidence is checked against min_conf before checking against the labels. There are a lot of objects detected in a room: tv, chair, table, ... so the confidences were all checked before they're all discarded due to not matching the label filter. I guess these are simple calculation, doesn't affect performance much, maybe low priority.

I haven't profiled that part yet, if I remember correctly this was needed due to how the defined zones filtering comes into play. If it makes sense to change it (and it doesn't break anything) I am not against it.

**Comment by hqhoang** [comment]
> Can you post more of the logs previous to the exception after pulling this next commit. I am adding a few debug statements when the offending vars are set.
> 
> Are you using API token auth?
Not sure if I'm using token auth, I think I am. Pasting the whole log, hopefully you can find what you're looking for.

```
noname@GTX1050:/var/lib/zmeventnotification$ sudo -u www-data python3 bin/zm_detect.py --eventid 308808 --debug 

01/26/22 22:43:35.082913 zmesdetect_m13[8887] DBG1 new_yaml.py:282->[config:init: the supplied config file exists -> '/etc/zm/objectconfig.yml']

01/26/22 22:43:36.243239 zmesdetect_m13[8887] DBG1 api.py:182->[Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/pyzm/api.py", line 174, in cred_dump
    "access_token_datetime": self.access_token_datetime.timestamp(),
AttributeError: 'NoneType' object has no attribute 'timestamp'
]

01/26/22 22:43:36.242268 zmesdetect_m13[8887] DBG1 api.py:181->[CRED DUMP DEBUG>>>  exception as str -> 'NoneType' object has no attribute 'timestamp']

01/26/22 22:43:36.241339 zmesdetect_m13[8887] ERR api.py:180->[api: ERROR while attempting to dump current credentials]

01/26/22 22:43:35.096638 zmesdetect_m13[8887] DBG1 new_yaml.py:359->[config:zmes: successfully replaced 17 secrets in the base config -> ['ML_USER', 'ZM_API_PORTAL', 'MQTT_USERNAME', 'PUSHOVER_APP_TOKEN', 'mlapi_one_key', 'HA_SERVER', 'ML_PASSWORD', 'PUSHOVER_USER', 'PUSHOVER_USER_KEY', 'ZM_USER', 'PUSHOVER_PASS', 'HA_TOKEN', 'ZM_PORTAL', 'MQTT_PASSWORD', 'ZM_PASSWORD', 'PUSHOVER_ERR_TOKEN', 'PLATEREC_ALPR_KEY']]

01/26/22 22:43:35.097701 zmesdetect_m13[8887] DBG1 new_yaml.py:867->[zmes:api create: building ZM API Session]

01/26/22 22:43:35.885919 zmesdetect_m13[8887] DBG1 zm_detect.py:825->[zmes: ZM DB SET GLOBAL MONITOR ID!]
01/26/22 22:43:35.953863 LOG_BUFFER[8887] DBG1 new_yaml.py:300->[config:init: default configuration built (no secrets or substitution vars replaced, yet!)]

01/26/22 22:43:36.277855 zmesdetect_m13[8887] WAR zm_detect.py:1046->[zmes: there was an error during the remote detection! -> 'NoneType' object does not support item assignment]01/26/22 22:43:35.955349 LOG_BUFFER[8887] DBG1 new_yaml.py:208->[config:zmes:proc: 34 built in default keys added to the 'base' config -> ['force_mpd', 'same_model_high_conf', 'basic_user', 'basic_password', 'image_path', 'max_detection_size', 'contained_area', 'delete_after_analyze', 'write_debug_image', 'write_image_to_zm', 'hass_enabled', 'hass_people', 'hass_notify', 'hass_cooldown', 'push_err_device', 'push_jpg', 'push_jpg_key', 'push_gif', 'push_gif_key', 'push_debug_device', 'push_cooldown', 'mqtt_topic', 'mqtt_broker', 'mqtt_port', 'tls_cert', 'tls_key', 'fp16_target', 'alpr_api_type', 'alpr_url', 'openalpr_recognize_vehicle', 'openalpr_country', 'openalpr_state', 'openalpr_min_confidence', 'smart_fs_thresh']]
01/26/22 22:43:35.956592 LOG_BUFFER[8887] DBG1 new_yaml.py:143->[config:zmes:proc: substituting '{{variables}}' for the 'base' config keys]
01/26/22 22:43:35.958633 LOG_BUFFER[8887] DBG1 new_yaml.py:165->[config:zmes: successfully replaced 5 default sub vars in the base config -> ['base_data_path', 'coral_models', 'tinyyolo_models', 'yolo4_models', 'yolo3_models']]
01/26/22 22:43:35.959172 LOG_BUFFER[8887] DBG1 zm_detect.py:829->[perf:ZM DB: time to grab Monitor ID (13): 0.1322]

01/26/22 22:43:35.960401 LOG_BUFFER[8887] DBG1 new_yaml.py:326->[config:zmes: starting '{[secrets]}' substitution]
01/26/22 22:43:35.966803 LOG_BUFFER[8887] DBG1 new_yaml.py:364->[config:zmes: there are 4 secrets configured that have no substitution candidates in the base config or the secrets file -> ['AWS_ACCESS_KEY_ID', 'ZM_ML_USER', 'ZM_ML_PASSWORD', 'AWS_SECRET_ACCESS_KEY']]
01/26/22 22:43:35.968609 LOG_BUFFER[8887] DBG1 new_yaml.py:392->[config:zmes: starting '{{variable}}' substitution]
01/26/22 22:43:35.973458 LOG_BUFFER[8887] DBG1 new_yaml.py:413->[config:zmes: successfully replaced 51 sub vars in the base config -> ['face_train_model', 'yolo4_object_processor', 'known_images_path', 'yolo4_object_framework', 'openalpr_cmdline_binary', 'tpu_object_labels', 'alpr_detection_pattern', 'same_model_sequence_strategy', 'object_min_confidence', 'tpu_object_weights_mobiledet', 'tpu_efficientdet_lite3', 'save_unknown_faces_leeway_pixels', 'frame_set', 'unknown_images_path', 'yolo4_object_weights', 'openalpr_cmdline_params', 'face_detection_framework', 'model_sequence', 'tpu_tf2_mobilenetv2', 'smart_fps_thresh', 'tpu_face_weights_mobilenetv2', 'face_upsample_times', 'yolo4_object_labels', 'platerec_min_dscore', 'alpr_key', 'save_unknown_faces', 'face_model', 'tpu_max_lock_wait', 'face_detection_pattern', 'face_num_jitters', 'contained_area', 'object_detection_pattern', 'cpu_max_processes', 'tpu_object_weights_mobilenetv2', 'show_models', 'past_det_max_diff_area', 'tpu_max_processes', 'yolo4_object_config', 'tpu_min_confidence', 'platerec_stats', 'frame_strategy', 'max_detection_size', 'gpu_max_processes', 'face_recog_dist_threshold', 'gpu_max_lock_wait', 'match_past_detections', 'platerec_min_score', 'cpu_max_lock_wait', 'openalpr_cmdline_min_confidence', 'disable_locks', 'tpu_object_framework']]
01/26/22 22:43:35.973933 LOG_BUFFER[8887] DBG1 new_yaml.py:417->[config:zmes: there was 1 sub var configured that has no substitution in the base config -> ['object_confidence_upper']]
01/26/22 22:43:35.976184 LOG_BUFFER[8887] DBG1 new_yaml.py:219->[config:zmes:proc: Base config has been built, all properly configured {[secrets]} and {{vars}} have been replaced! Remember you must manually build per monitor overrode configurations]
01/26/22 22:43:35.976608 LOG_BUFFER[8887] DBG1 zm_detect.py:896->[perf:zmes: building the initial config took 0.1475261199993838]
01/26/22 22:43:35.981923 LOG_BUFFER[8887] DBG2 api.py:124->[api: SSL certificate verification disabled (encryption enabled, vulnerable to MITM attacks)]
01/26/22 22:43:35.982663 LOG_BUFFER[8887] DBG1 api.py:311->[api: no token found, trying user/pass for login]
01/26/22 22:43:35.989474 LOG_BUFFER[8887] DBG2 api.py:348->[api: detected API ver 2.0+, using token system]
01/26/22 22:43:36.010793 LOG_BUFFER[8887] DBG1 new_yaml.py:510->[config:zmes:13: no secrets were replaced during overrode config build]
01/26/22 22:43:36.067904 LOG_BUFFER[8887] DBG1 new_yaml.py:909->[zmes:api create: ZM API created]
01/26/22 22:43:36.106206 LOG_BUFFER[8887] INF zm_detect.py:902->[------|  FORKED NEO --- app->Hooks: 7.0.0 - pyzm: 0.4.0 - ES: 7.0.0 - OpenCV:4.5.2 |------]
01/26/22 22:43:36.106578 LOG_BUFFER[8887] DBG1 zm_detect.py:933->[zmes: Monitor ID (13) has an overrode config, switching to it!]
01/26/22 22:43:36.106836 LOG_BUFFER[8887] DBG1 new_yaml.py:454->[config:zmes:13: attempting to build an overrode config from monitor '13' overrides]
01/26/22 22:43:36.107592 LOG_BUFFER[8887] DBG1 new_yaml.py:469->[config:zmes:13: monitor '13' overrides, starting '{[secrets]}' substitution]
01/26/22 22:43:36.108223 LOG_BUFFER[8887] DBG1 new_yaml.py:522->[config:zmes:13: this monitor has a key defined that is not in the BASE config?! adding 'person_max_detection_size'=65% to the BASE config]
01/26/22 22:43:36.108505 LOG_BUFFER[8887] DBG1 new_yaml.py:522->[config:zmes:13: this monitor has a key defined that is not in the BASE config?! adding 'person_min_confidence'=40% to the BASE config]
01/26/22 22:43:36.108783 LOG_BUFFER[8887] DBG1 new_yaml.py:522->[config:zmes:13: this monitor has a key defined that is not in the BASE config?! adding 'car_past_det_max_diff_area'=45% to the BASE config]
01/26/22 22:43:36.109059 LOG_BUFFER[8887] DBG1 new_yaml.py:522->[config:zmes:13: this monitor has a key defined that is not in the BASE config?! adding 'object_confidence_upper'=79% to the BASE config]
01/26/22 22:43:36.109488 LOG_BUFFER[8887] DBG1 new_yaml.py:560->[config:zmes:13: no vars were replace during overrode config build]
01/26/22 22:43:36.128767 LOG_BUFFER[8887] DBG1 new_yaml.py:708->[config:zmes:13: 5 keys overridden in the 'base' config -> ['object_detection_pattern', 'frame_set', 'model_sequence', 'match_past_detections', 'past_det_max_diff_area']]
01/26/22 22:43:36.129245 LOG_BUFFER[8887] DBG1 new_yaml.py:713->[config:zmes:13: 4 keys that did not have a 'base' value to override that are now in the 'base' config -> ['person_max_detection_size', 'person_min_confidence', 'car_past_det_max_diff_area', 'object_confidence_upper']]
01/26/22 22:43:36.129531 LOG_BUFFER[8887] DBG1 new_yaml.py:718->[config:zmes:13: re-parsing SUB VARS...]
01/26/22 22:43:36.129814 LOG_BUFFER[8887] DBG1 new_yaml.py:392->[config:zmes: starting '{{variable}}' substitution]
01/26/22 22:43:36.130411 LOG_BUFFER[8887] DBG1 new_yaml.py:413->[config:zmes: successfully replaced 1 sub var in the base config -> ['object_confidence_upper']]
01/26/22 22:43:36.132121 LOG_BUFFER[8887] DBG1 new_yaml.py:720->[self.config.get('ml_sequence', {}).get('general') = {'model_sequence': 'object', 'disable_locks': 'False', 'match_past_detections': 'True', 'past_det_max_diff_area': '5%', 'aliases': {'vehicles': ['car', 'bus', 'truck', 'boat', 'motorcycle'], 'plants': ['broccoli', 'pottedplant', 'potted_plant'], 'animals': ['dog', 'cat', 'mouse', 'horse']}}]
01/26/22 22:43:36.132437 LOG_BUFFER[8887] DBG1 new_yaml.py:723->[config:zmes:13: overwriting old per monitor global config as new data is available!]
01/26/22 22:43:36.133056 LOG_BUFFER[8887] INF zm_detect.py:945->[zmes: Setting up signal handlers for log 'rotation' and 'interrupt']
01/26/22 22:43:36.215446 LOG_BUFFER[8887] DBG1 zm_detect.py:985->[zmes: this is a 'PAST' (debugging?) event!]
01/26/22 22:43:36.216043 LOG_BUFFER[8887] INF zm_detect.py:395->[|----------= Encrypted Route Name: 'mlapi_one' | Gateway URL: 'http://localhost:5000/api/v1' | Weight: 0 =----------|]
01/26/22 22:43:36.217755 LOG_BUFFER[8887] DBG1 zm_detect.py:337->[zmes:mlapi:login: No need to login, access token is valid for 2712.1710953712463 sec]
01/26/22 22:43:36.238099 LOG_BUFFER[8887] INF ZMLog.py:340->[Connected to ZoneMinder Logging system with user 'www-data' -> '/var/log/zm/zmesdetect_m13.log']
01/26/22 22:43:36.240307 LOG_BUFFER[8887] DBG1 api.py:167->[
ASKING FOR A CRED DUMP NOW!!
]
01/26/22 22:43:36.278512 LOG_BUFFER[8887] DBG1 zm_detect.py:1047->[Traceback (most recent call last):
  File "bin/zm_detect.py", line 1014, in main_handler
    remote_response = remote_detect(options=stream_options, args=args, route=route)
  File "bin/zm_detect.py", line 503, in remote_detect
    kickstart['api_url'] = g.api.api_url
TypeError: 'NoneType' object does not support item assignment
]
01/26/22 22:43:36.279129 LOG_BUFFER[8887] DBG1 zm_detect.py:1908->[zmes: no predictions returned from detections]
01/26/22 22:43:36.279804 LOG_BUFFER[8887] DBG1 zm_detect.py:1967->[perf:zmes:FINAL: 'Monitor': Driveway (13)->'Event': 308808 ['PAST' event] [total:0.45273770300264005] [pre detection:0.3890618680015905][detection:0.06182116999843856] [after detection: 0.0005837780008732807] ]

01/26/22 22:43:35.959997 zmesdetect_m13[8887] DBG1 new_yaml.py:320->[config:zmes: the configured secrets file exists and is a file -> '/etc/zm/zm_secrets.yml']

```

**Comment by baudneo** [comment]
That's bizarre. For token auth, are you using user/password authentication to authenticate to zoneminder? Or is auth turned off. I am away from my system ATM and can not test if that's the issue.

I have a couple.ofnideas as to what's going on but need a bit more info.

**Comment by baudneo** [comment]
Ok so I said intoy system and tested with auth turned off and am getting those errors. I won't be able to do proper fix until tmrw when I am at home. 

For the time being enable auth 9n your zm web gui->options>system and create a user and password to login to zm. This will allow you to use mlapi for tonight if you want or wait until tmrw to get the fix for a no auth instance.

**Comment by hqhoang** [comment]
ah, yes, I do have a user name and password from the beginning but I disabled "opt_use_auth" as the server isn't exposed to outside, and it's a hassle to login everytime I use the web interface.

Turning on auth indeed eliminated the error and went through the mlapi on port 5000. I just need to copy the settings from objectconfig.yml over to mlapiconfig.yml. Thanks.

**Comment by baudneo** [comment]
I am back in the homelab and will be working on no auth mechanism, hopefully by end of day it should be implemented. I should also work on basic auth while I'm there lol.

**Comment by hqhoang** [comment]
while you're at it, a minor error-prevention:

```
@@ -336,7 +336,7 @@ class ZMLog:
         self.is__active = True
         show_log_name = f"'{self.log_filename}'"
         if ZM_INSTALLED:
-            if self.buffer:
+            if self.buffer is not None and self.buffer:
                 self.buffer.info(
                     f"Connected to ZoneMinder Logging system with user '{getuser()}'"
                     f"{f' -> {show_log_name}' if self.log_file_handler else ''}"
```
I'm still trying to figure out why an event started in ZM, but only trigger mlapi 3-4 seconds later. The monitor buffer currently has 0 warmup frames, 10 pre/post image count.

**Comment by baudneo** [comment]
Enable the 'sanitize_logs' option and post the log output for both ZMES and mlapi. I have reworked a bunch of stuff and am testing that codebase. My current test code base logs will be very different than the current good_enough and pull_req branches.

I haven't had that issue as of yet, usually when a detection is run, it is immediately passed to mlapi.

**Comment by baudneo** [comment]
Also, how do you store the events? are you using mp4 writer and passthrough or JPEG storage? JPEG storage is MUCH faster when running detections as ZM does not need to grab frames from the mp4 container.

**Comment by hqhoang** [comment]
I don't use a video writer (value is "disabled" for all monitors). I only save JPEG (frames + analysis) as I have a standalone NVR that records all cameras on motion detection. I'm using ZM only for detection. The average load of my laptop with 4 active monitors is usually 04-0.6.

Maybe I misunderstood how to read the times. I confirmed that the clock on the laptop is the same as my phone, down to the seconds, then intentionally ran into the room at 14:48:00. On ZM's Events page, it shows the event started at 14:48:00:
```
309305 | Event- 309305 | Lower Middle | Motion [20] :detected:person(77%-yolo[GPU]) Motion: All | 01/30 14:48:00 until 01/30 14:48:09 | 00:00:08 | 128 | 58 | 58 | 1 | 1
```
and /var/log/zm/zmesdetect_m4.log show it doesn't call zm_detect.py until 14:48:04
```
01/30/22 14:48:04.893051 zmesdetect_m4[92505] INF zm_detect.py:902 [------|  FORKED NEO --- app->Hooks: 7.0.0 - pyzm: 0.4.0 - ES: 7.0.0 - OpenCV:4.5.2 |------]
01/30/22 14:48:04.920806 zmesdetect_m4[92505] INF zm_detect.py:945 [zmes: Setting up signal handlers for log 'rotation' and 'interrupt']
01/30/22 14:48:04.953559 zmesdetect_m4[92505] INF zm_detect.py:395 [|----------= Encrypted Route Name: 'mlapi_one' | Gateway URL: 'http://localhost:5000/api/v1' | Weight: 0 =----------|]
01/30/22 14:48:06.008308 zmesdetect_m4[92505] INF zm_detect:1151 [zmes:prediction: '[20] person(77%-yolo[GPU])']

```
I added  "date >> /tmp/zm_event_start.log"  to /var/lib/zmeventnotification/bin/zm_event_start.sh, and the log file showed "14:48:03", so the majority of the delay is likely before calling zm_event_start.sh.

I confirmed with a few more tests, and there's consistently a delay of approx 3 seconds between the event start and the trigger of zm_event_start.sh.

**Comment by baudneo** [comment]
> and /var/log/zm/zmesdetect_m4.log show it doesn't call zm_detect.py until 14:48:04

This is due to the design of the system, in the Perl config file there is an option named `event_check_interval:`. By default it is set to 5 seconds, this means that ZMES will scan the SHM every 5 seconds looking for any new events. 

So in your case lets assume that ZMES scans the SHM every 5 seconds evenly. 14:47:59 it is scanned and it does not see the new event as it has not happened yet. The event is created at 14:48:00 and starts doing its work. ZMES scans SHM at 14:48:04 and finds the new event and then starts its work.

The ZM dev team has created a hook that can run an external script in the 1.37 master branch but in my personal testing ZMES SHM scans are usually faster around 90% of the time. I will include a screenshot of where the new hook script is set up. This is the 'source' modal when you are setting the monitors buffers, name, etc.

![zmes-new-hook](https://user-images.githubusercontent.com/86508179/151718630-7ad23dc1-5429-4acd-b78c-5c4ef8af56d9.png)

**Comment by baudneo** [comment]
I have a snippet of code in `zm_detect.py` to catch the --new CLI argument. It will log the time it was called in the monitors zm_detectm*.log and exit so you can compare when ZMES kicks off on an event and when the new external `EventStartCommand` kicks off. The --new CLI catcher should be line 878 in `zm_detect.py` master branch.

Here is a script you can put in `/var/lib/zmeventnotification/bin/test.sh` to test the timing of the new ZM eventstartcommand ->
```
#!/usr/bin/env bash
trap 'cleanup' SIGINT SIGTERM
cleanup() {
   # Don't echo anything here
   exit 1
}
#

CONFIG_FILE="/etc/zm/objectconfig.yml"
ZMES_DIR="/var/lib/zmeventnotification"

#MID=''
#ETYPE=''
EID=$1
#for arg in "$@"
#do
#    case $arg in
#        -m=*|--monitor-id=*)
#            MID="${arg#*=}"
#            shift # past argument=value
#            ;;
#        -et=*|--event-type=*)
#            ETYPE="${arg#*=}"
#            shift # past argument=value
#            ;;
#    esac
#done
#
##[[ ! -e "${ZMES_DIR}/bin/zm_detect.py" ]] && echo "${ZMES_DIR}/bin/zm_detect.py does not exist!" && exit 1
DET_SCRIPT=("${ZMES_DIR}/bin/zm_detect.py" --eventid "${EID}" --config "${CONFIG_FILE}" --live --debug --event-type "start" --new)
DET_OUTPUT=$("${DET_SCRIPT[@]}")
echo 0
exit 0
```

Make sure you `chmod +x` the new script and then add its absolute path to the `EventStartCommand` textbox `/var/lib/zmeventnotification/bin/test.sh`

**Comment by hqhoang** [comment]
> This is due to the design of the system, in the Perl config file there is an option named `event_check_interval:`. By default it is set to 5 seconds, this means that ZMES will scan the SHM every 5 seconds looking for any new events. **BE AWARE** - for some reason I have tried lowering this setting and it still scans every 5 seconds regardless, I am trying to hunt down the bug there. It is a weird one, I set the hard coded default to 2 seconds and set that option to 2 seconds and still ZMES scans every 5 seconds. So, _changing that option may not make ZMES scan SHM less than every 5 seconds_.

Thanks much, that explains it. Periodically polling of course causes random delay between 0 and max of the scanning window. At least the issue has been looked into, I'll wait for further development/improvement on this front. Let me know if there's anything I can help (testing) with.

**Comment by baudneo** [comment]
I am rerunning the `EventStartCommand` timing tests and for some reason, `EventStartCommand` is faster now basically 100% of the time. I'll keep testing and see what's going on with that as the results are boggling my mind. The only thing I can think is if I was previously testing on a monitor using mp4 storage instead of JPEG but still that doesn't make sense.

Same event
```

  ZONEMINDER: EventStartCommand was called -> 01/30/22 15:58:39.473912

  THIS IS A ZMES SHM SCAN EVENT 01/30/22 15:58:46.570355
```

**Comment by hqhoang** [comment]
That's good news. I'll see if I can install the master branch in a parallel setup without messing up the current ZM.

**Comment by baudneo** [comment]
Give me an hour to fix the mess I made in the repos for timing. Or in that example.test.sh script remove the `--new` flag.

There seems to be a large performance boost using `EventStartCommand`. It also causes some issues with frame buffer not being populated due to the detection kicking off so fast.

**Comment by hqhoang** [comment]
> It also causes some issues with frame buffer not being populated due to the detection kicking off so fast.

Is it the problem with the first frame not being available for detection? If so, I guess the codes can just trigger/flag once the first frame is written (or providing direct access if not so currently).

If the problem is with subsequent frames not being available quick enough, if we know the fps of the monitor/camera, given that we know how long it takes to perform ML on the last frame, we can calculate what the next available frame is. Perhaps we could introduce an "adaptive_interval" option for the frameset. Either the user explicitly chooses this option for `frameset`, or it's an internal mechanism that would override the user's frameset if it detects what the user specified won't be available yet. Given that we already have  `contig_frames_before_error` and `delay_between_attempts`, it's probably better to have this "adaptive_interval" as an option for `frameset`, users won't specify the frame IDs with this option. 

This "adaptive_interval" has an advantage over a fixed frameset as it can dynamically adjust based on the system load, e.g. one event going on or multiple events queuing up. It can even have fine-tune options, such as `max_jump` so that it can fall behind the latest available frame a bit to not miss the actions in between.

I'm sorry if I opened a can of worms, you make the call, it's easier said than done as I don't know enough to do the coding yet :-)

**Comment by baudneo** [comment]
There is a plan to make what you are describing, currently FPS is calculated and there is a dynamic sleep function that is based on 'fps_thresh' option. If you are using EventStartCommand.sh I recommend adding delay: 2.0 into 'stream_sequence' to allow the frame buffer to get some frames into it if you are experiencing issues due to the detections kicking off too fast.
